draft post about record-known-result (#61)
pypy/pypy.org: posts/2022/06/record-known-result.rst

==================================================================================
A Hint for Language-Specific Runtime Function Optimization in RPython's Meta-JIT
==================================================================================

RPython's meta-JIT cannot reason about the properties of many RPython
implementation-level functions, because those are often specific to the
different interpreters written in RPython. Expressing language-specific
optimizations on these functions would allow the JIT to more effectively reason
about language-specific data structures and potentially remove costly calls
into the runtime support code. In this blog post I will describe a new hint
that makes it possible to express invariants about the functions of the data
structures used in a specific language interpreter. These hints make it
possible for the meta-JIT optimizer to do more language-specific optimizations
without having to change the meta-JIT itself.

Introduction and Background
===========================

RPython has a meta-JIT, that can be applied to a variety of languages, one of
them being PyPy, a Python implementation (that's really two implementations,

> one PyPy2, one PyPy3). Most modern languages have large runtimes, meaning many

Reviewer: ```suggestion
one PyPy2, one PyPy3). Most modern languages have many
```

built-in functions and data-types that are written in the implementation
language (RPython, in our case, C in the case of CPython). We want the JIT

> optimizer to be able to reason about functions that are part of the runtime
> support that is defined in the interpreter.

Reviewer: ```suggestion
optimizer to be able to reason about these functions and datatypes.
```

One way to do this would be to extend the meta-JIT to able to do

> language-specific rewrites on these functions, effectively adding a new

Reviewer: ```suggestion
language-specific rewrites, effectively adding a new
```

Author: I'll do this even shorter, I think, removing the part of the sentence entirely.

optimization pass that knows how to deal with these functions. This approach
however goes against the idea of a meta-JIT, since we want the JIT to be
language independent. So we developed various tools in the past that make it
possible for the author of RPython interpreters to express various

> optimizations. One of the very important ones that has been there forever is

Reviewer: ```suggestion
optimizations. Since RPython is written in python, we can write these
hints to the JIT as decorators. For instance, one hint that has been there forever is
```

Author: I added a bit more text to be explicit about what I mean by "hints"

by the previous result.

.. __: https://en.wikipedia.org/wiki/Common_subexpression_elimination

There are lots of examples of ``@elidable`` functions, both in RPython in
general and in the PyPy interpreter. In the post linked above they are used for
optimizing object operations in a tiny interpreter.

But also a lot of functionality that is more directly exposed to the Python
programmer is elidable. Most string methods are elidable for example:
``str.lower`` will always return the same result for some argument, and if you
call it twice on the same string, it will also return the same result. So
adding the ``@elidable`` decorator to the implementation of the string method
allows the JIT to constant-fold it if it's applied to a constant string, and to
remove a second identical call. Another big class of examples are all the
implementation functions for our big integer operations, which underlie the
Python ``int`` type once the values don't fit into a machine word anymore. On
the implementation level we implement those in an ``rbigint`` class, and most
methods of that are also elidable. This enables the JIT to constant-fold big
integer addition, and do CSE on big integer arithmetic.

> This is all very useful! But it's still only a limited amount of things that
> the interpreter author can express to the JIT with this one function decorator.

Reviewer: Do we have a list of hints we could link to?
```suggestion
This is all very useful! There are additional hints 
the interpreter author can express to the JIT.
```

Author: I was trying to look in our docs, but seems the answer unfortunately is "no, not at all" :-(
(and in general the docs are quite bad around the jit)

Recently I merged a branch that adds two new hints at the interpreter author

> can use to communicate possible optimizations to the meta-JIT! The work has
> been happening on and off in the last three years, the scope was a bit too big
> and some parts were not done and I didn't seem to converge either. So I
> recently decided that this was going on for too long already and to merge a
> minimal version of the branch, which contains only the support for the two new
> hints, but not really many interesting *applications* for the hints to the PyPy
> interpreter. Those can be added in a much more localized fashion over time.

Reviewer: ```suggestion
can use to communicate possible optimizations to the meta-JIT! The work has been
a long-term project of mine. Currently support for these hints has been merged. In future enhancements I will apply them where appropriate.
```

Author: fixed this slightly differently

``record_known_result``
=======================

So, what are the hints? One of them is called ``record_known_result``, and that

> hint is what this blog post is about. The one one of them is called

Reviewer: ```suggestion
hint is what this blog post is about. The other is called
```

``record_exact_value``, it's conceptually quite simple, but it's much harder to
see how it can be used. It was implemented by Lin Cheng from Cornell, and it is

> described (together with a possible optimization) in another paper__ – I will
> talk about that one in a later post, if I get to it.

Reviewer: ```suggestion
described (together with a possible optimization) in another paper__.
```

each other, so that ``f(g(x)) == x``  for all ``x`` (example: negation on
numbers is its own inverse, ``--x == x``). Sometimes functions are
*idempotent__*, which means that if you call the function several times you can
remove all but the first call. An example would be ``abs`` on numbers, after
the first call to ``abs`` the result is positive, so calling the function again
on the result has no effect, i.e. ``abs(abs(x)) == abs(x)``. These properties
could in theory be nicely used by a hypothetical optimizer that knows about
them and the functions. However, as described above, we don't want to change
the meta-JIT to add knowledge about interpreter-specific functionality. So we
wanted to add a hint that can express them to the meta-JIT.

.. __: https://en.wikipedia.org/wiki/Idempotence#Idempotent_functions

What could hints look like, that make it possible to express these
properties? At first I was experimenting with some declarative decorators
like ``@idempotent`` and ``@is_inverse_of(func)`` but I felt like it wouldn't
scale to add lots of these decorators and support for all of them in the
meta-JIT. In the end I found a not fully obvious hint that is not a

> decorator, that is powerful enough to implement at least these last two and a
> whole bunch of other patterns. This hint piggy-backs on the existing CSE

Reviewer: ```suggestion
decorator, that is powerful enough to implement at least these last two and
many similar patterns. This hint piggy-backs on the existing CSE
```

function in the runtime that is responsible for computing the negative value
of a big integers? The implementation of ``rbigint_neg`` looks roughly like
this (it's actually a method and a tiny bit more complicated, but not really
significantly so):

.. code:: python

    @elidable
    def rbigint_neg(self):
        return rbigint(digits=self.digits, sign=-self.sign)

If we want to use the new hint to express that ``rbigint_neg(rbigint_neg(x)) ==
x``, we need to rewrite the function somewhat, by introducing a pure helper
function that does the actual computation, and turning the original function
into a wrapper that calls the helper:

.. code:: python

    @elidable
    def _rbigint_neg_helper(self):
        return rbigint(digits=self.digits, sign=-self.sign)
        
    def rbigint_neg(self):
        res = _rbigint_neg_helper(self)
        record_known_result(self, _rbigint_neg_helper, res)
        return res

``record_known_result`` is a new function in the ``rpython.rlib.jit`` library that
has the signature ``record_known_result(result, function, *args)``. What does
this function do? Outside of the JIT, a call to that function is simply

> ignored. But when we trace the ``rbigint_neg`` function the hint tells the JIT

Reviewer: ```suggestion
ignored. But when we trace the ``rbigint_neg`` function, the hint tells the JIT
```

with ``res`` (because ``res`` is already all lower cased, as its the result of
a call to ``lower``), i.e. that ``_bytes_lower_helper`` is idempotent. (There
are also other properties of lower and upper we could express in this way, for
example that ``bytes.lower(bytes.upper(x)) == bytes.lower(x)``, let's leave it
at that for now though).

Both of these usage patterns of ``record_known_result`` could of course also be
pulled out into general decorators again. For example a generic ``@idempotent``
decorator could be implemented like this:

.. code:: python

    def idempotent(func):
        func = elidable(func) # idempotent implies elidable
        def wrapper(arg):
            res = func(arg)
            record_known_result(res, func, res)
            return res
        return wrapper

Then the decorator could be used like this for ``bytes_lower``:

.. code:: python

    @idempotent
    def bytes_lower(b):
        # implementation as in the original code above
        ...

Implementing ``record_known_result``
========================================

How is ``record_known_result`` implemented? As I wrote above, the implementation

> of that hint is piggy-backing on the existing support for ``elidable`` functions

Reviewer: ```suggestion
of that hint builds on the existing support for ``elidable`` functions
```

hint is only a new tool in the interpreter author toolbox, and we still need to
find the best places to use this tool. The only use of the hint so far is an
annotation that tells the JIT that encoding and decoding to and from utf8 are
inverses of each other, to be able to optimize this kind of code:
``x = someunicode.encode("utf-8").decode("utf-8")`` by replacing ``x`` with
``someunicode`` (of course in practice there is usually again some distance
between the encode and decode calls). This happens in a bunch of places in real
code that I saw, but I didn't do a super careful study of what the performance
effect is yet.

Limitations
=============

What are the problems and the limitations of the approach I described in this
post?

Correctness remains tricky! If you write the wrong hints, the meta-JIT will
potentially miscompile your users' programs. To at least get some signal for
that, ``record_known_result`` actually performs the hinted call and does an
assert on the result if you run the program untranslated while executing tests.
In combination with for example property-based testing this can find a lot of
the bugs, but is of course no guarantee.

> Many things aren't expressible! A lot less powerful than some of the recent
> pattern based optimization systems that allow the non-compiler authors to
> express rewrites. Instead, we designed the hint to minimally fit into the

Reviewer: ```suggestion
Many things aren't expressible! The new hint is much less powerful than
some of the recent
pattern based optimization systems that allow the non-compiler authors to
express rewrites. Instead, we designed the hint to minimally fit into the
```

allows the JIT to constant-fold them if they are applied to a constant string, and to
remove a second identical call.

Another big class of examples are all the
implementation functions for our big integer operations, which underlie the
Python ``int`` type once the values don't fit into a machine word anymore. On
the implementation level we implement those in an ``rbigint`` class, and most
methods of that are also elidable. This enables the JIT to constant-fold big
integer addition, and do CSE on big integer arithmetic.

Limitations of ``@elidable``
-------------------------------

This is all very useful! But it's still only a limited amount of things that
the interpreter author can express to the JIT with this one function decorator.
Recently I merged a branch that adds two new hints that the interpreter author
can use to communicate possible optimizations to the meta-JIT! The work has
been an ongoing project since a while. So far, only the meta-JIT support has
been merged, in future enhancements we plan to apply them to the PyPy Python
interpreter where appropriate.

``record_known_result``
=======================

> So, what are the hints? One of them is called ``record_known_result``, and that

Reviewer: ```suggestion
So, what are the new hints? One of them is called ``record_known_result``, and that
```

In the case of function inverses, dead code elimination also plays an important
role. Let's look at the trace of a double negation, maybe like this:
``x = -y; ...; print(-x)``:

.. code:: python

    r1 = call_elidable((_rbigint_neg_helper), a1)
    record_known_result(a1, (_rbigint_neg_helper), r1)
    ... intermediate stuff
    r2 = call_elidable((_rbigint_neg_helper), r1)
    record_known_result(r1, (_rbigint_neg_helper), r2)
    print(r2)

After CSE, the second call is removed and the trace looks like this, because
``r2`` was found to be the same as ``a1``:

.. code:: python

    r1 = call_elidable((_rbigint_neg_helper), a1) # dead
    ... intermediate stuff, CSEd
    # call removed
    print(a1)

Now dead code elimination notices that the first call is not needed any more
either and removes it.

What is good about this design? It very neatly ties into the existing
infrastructure and is basically only about 100 lines of changes in the
meta-JIT. The amount of work the optimizer does stays essentially the same, as
the new hints are basically directly usable by CSE which we already do anyway.

Performance effects
====================

> So far, we haven't actually used this new hint in PyPy much! At this point, the

Reviewer: ```suggestion
So far, we haven't actually used this new hint in PyPy much. At this point, the
```

hint is only a new tool in the interpreter author toolbox, and we still need to
find the best places to use this tool. The only use of the hint so far is an
annotation that tells the JIT that encoding and decoding to and from utf8 are
inverses of each other, to be able to optimize this kind of code:
``x = someunicode.encode("utf-8").decode("utf-8")`` by replacing ``x`` with
``someunicode`` (of course in practice there is usually again some distance
between the encode and decode calls). This happens in a bunch of places in real
code that I saw, but I didn't do a super careful study of what the performance
effect is yet.

Limitations
=============

What are the problems and the limitations of the approach I described in this
post?

Correctness remains tricky! If you write the wrong hints, the meta-JIT will
potentially miscompile your users' programs. To at least get some signal for
that, ``record_known_result`` actually performs the hinted call and does an
assert on the result if you run the program untranslated while executing tests.
In combination with for example property-based testing this can find a lot of
the bugs, but is of course no guarantee.

> Many things aren't expressible! The new hint is much less powerful than some of

Reviewer: ```suggestion
Many things aren't expressible. The new hint is much less powerful than some of
```

the recent pattern based optimization systems (e.g. `metatheory.jl`__) that
allow library authors to
express rewrites. Instead, we designed the hint to minimally fit into the
existing optimizers at the cost of power and ease of use. The most
obvious limitation compared to pattern based approaches is that the

> ``record_known_result`` hint cannot quantify over unknown values, only use once

Reviewer: ```suggestion
``record_known_result`` hint cannot quantify over unknown values. It can only use values
```
